include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -0,0 +1,371 @@
+ * underlying LAPACK implementation supports only float and double numbers, so
+ * only these two types are currently supported.
+ *
+ * @tparam size Compile-time array lenghts. By default at -1, which means that
+ * the run-time info stored in the matrices passed to the reinit()
+ * function is used.
+ *
+ * @author Martin Kronbichler, 2017
+ */
+template <int dim, typename Number, int size = -1>
+class TensorProductMatrix
+{
+public:
+  /**
+   * Constructor.
+   */
+  TensorProductMatrix();
+
+  /**
+   * Initalizes the matrix to the given mass matrix $M$ and derivative matrix
+   * $A$. Note that the current implementation requires $M$ to be symmetric
+   * and positive definite and $A$ to be symmetric.
+   */
+  void reinit (const FullMatrix<Number> &mass_matrix,
+               const FullMatrix<Number> &derivative_matrix);
+
+  /**
+   * Returns the number of rows of this matrix, given by the dim-th power of
+   * the size of the 1D matrices passed to the constructor.
+   */
+  unsigned int m() const;
+
+  /**
+   * Returns the number of columns of this matrix, given by the dim-th power
+   * of the size of the 1D matrices passed to the constructor.
+   */
+  unsigned int n() const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class.
+   */
+  void vmult (Vector<Number> &dst,
+              const Vector<Number> &src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class. Same as the other
+   * vmult function, but operating on plain pointers rather than a vector (no
+   * check of array bounds possible).
+   */
+  void vmult (Number *dst,
+              const Number *src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class.
+   */
+  void apply_inverse (Vector<Number> &dst,
+                      const Vector<Number> &src) const;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class. Same as the other
+   * vmult function, but operating on plain pointers rather than a vector (no
+   * check of array bounds possible).
+   */
+  void apply_inverse (Number *dst,
+                      const Number *src) const;
+
+private:
+  /**
+   * A copy of the @p mass_matrix object passed to the reinit() method.
+   */
+  FullMatrix<Number> mass_matrix;
+
+  /**
+   * A copy of the @p derivative_matrix object passed to the reinit() method.
+   */
+  FullMatrix<Number> derivative_matrix;
+
+  /**
+   * A vector containing the generalized eigenvalues of A s = lambda B s.
+   */
+  AlignedVector<Number> eigenvalues;
+
+  /**
+   * The matrix containing the generalized eigenvectors.
+   */
+  Table<2,Number> eigenvectors;
+
+  /**
+   * An array for temporary data.
+   */
+  mutable AlignedVector<Number> tmp_array;
```

Comment:
I see your point (and had actually thought about it). But I would like to avoid the mutex here because these are (very) small functions and thought the general use case to be a separate class for each thread. An alternative would be to have a thread local class variable for the temporary vector that we can query. How about that?